libsharedringbuffer/include/shared_ringbuffer/shared_ringbuffer.h

```diff
@@ -0,0 +1,233 @@
+#include <utils/util.h>
+#include <shared_ringbuffer/gen_config.h>
+
+/* Function pointer to be used to 'notify' components on either end of the shared memory */
+typedef void (*notify_fn)(void);
+
+/* Buffer descriptor */
+typedef struct buff_desc {
+    uintptr_t encoded_addr; /* encoded dma addresses */
+    unsigned int len; /* associated memory lengths */
+    void *cookie; /* index into client side metadata */
+} buff_desc_t;
+
+/* Circular buffer containing descriptors */
+typedef struct ring_buffer {
+    uint32_t write_idx;
+    uint32_t read_idx;
+    buff_desc_t buffers[CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT];
+} ring_buffer_t;
+
+/* A ring handle for enqueing/dequeuing into  */
+typedef struct ring_handle {
+    ring_buffer_t *used_ring;
+    ring_buffer_t *avail_ring;
+    /* Function to be used to signal that work is queued in the used_ring */
+    notify_fn notify;
+} ring_handle_t;
+
+
+/**
+ * Initialise the shared ring buffer.
+ *
+ * @param ring ring handle to use.
+ * @param avail pointer to available ring in shared memory.
+ * @param used pointer to 'used' ring in shared memory.
+ * @param notify function pointer used to notify the other user.
+ * @param buffer_init 1 indicates the read and write indices in shared memory need to be initialised.
+ *                    0 inidicates they do not. Only one side of the shared memory regions needs to do this. 
+ */
+void ring_init(ring_handle_t *ring, ring_buffer_t *avail, ring_buffer_t *used, notify_fn notify, int buffer_init);
+
+/**
+ * Check if the ring buffer is empty.
+ *
+ * @param ring ring buffer to check.
+ *
+ * @return true indicates the buffer is empty, false otherwise.
+ */
+static inline int ring_empty(ring_buffer_t *ring) 
+{
+    return !((ring->write_idx - ring->read_idx) % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT);
+}
+
+/**
+ * Check if the ring buffer is full
+ *
+ * @param ring ring buffer to check.
+ *
+ * @return true indicates the buffer is full, false otherwise.
+ */
+static inline int ring_full(ring_buffer_t *ring)
+{
+    return !((ring->write_idx - ring->read_idx + 1) % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT);
+}
+
+/**
+ * Notify the other user of changes to the shared ring buffers.
+ * 
+ * @param ring the ring handle used.
+ *
+ */
+static inline void notify(ring_handle_t *ring) 
+{
+    return ring->notify();
+}
+
+/**
+ * Enqueue an element to a ring buffer
+ *
+ * @param ring Ring buffer to enqueue into.
+ * @param buffer address into shared memory where data is stored.
+ * @param len length of data inside the buffer above.
+ * @param cookie optional pointer to data required on dequeueing. 
+ *
+ * @return -1 when ring is empty, 0 on success.
+ */
+static inline int enqueue(ring_buffer_t *ring, uintptr_t buffer, unsigned int len, void *cookie) 
+{
+    if (ring_full(ring)) {
+        ZF_LOGE("Ring full");
+        return -1; 
+    }
+
+    ring->buffers[ring->write_idx % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT].encoded_addr = buffer;
```

Comment:
Could the code redundancy a bit (even if the compiler might optimize this anyway)
```
buff_desc_t *buff_desc = &(ring->buffers[ring->write_idx % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT]);
buff_desc->encoded_addr = buffer
buff_desc->len = len;
buff_desc->cookie = cookie;
```